framework/src/modules/chain/dpos/delegates_list.js

```diff
@@ -53,33 +53,54 @@ class DelegatesList extends EventEmitter {
 		return shuffleDelegateListForRound(round, list);
 	}
 
-	async getDelegatePublicKeysSortedByVote() {
+	async getDelegatePublicKeysSortedByVoteWeight(tx) {
 		const filters = { isDelegate: true };
 		const options = {
 			limit: this.activeDelegates,
 			sort: ['voteWeight:desc', 'publicKey:asc'],
 		};
-		const accounts = await this.storage.entities.Account.get(filters, options);
+		const accounts = await this.storage.entities.Account.get(
+			filters,
+			options,
+			tx,
+		);
 		return accounts.map(account => account.publicKey);
 	}
 
 	async generateActiveDelegateList(round) {
-		let delegatePublicKeys = await this.storage.entities.RoundDelegates.getRoundDelegates(
+		const delegatePublicKeys = await this.storage.entities.RoundDelegates.getRoundDelegates(
 			round,
 		);
 
 		if (!delegatePublicKeys.length) {
-			delegatePublicKeys = await this.getDelegatePublicKeysSortedByVote();
-
-			await this.storage.entities.RoundDelegates.create({
-				round,
-				delegatePublicKeys,
-			});
+			throw new Error(`No delegate list found for round: ${round}`);
 		}
 
 		return delegatePublicKeys;
 	}
 
+	async createRoundDelegateList(round, tx) {
+		const delegatePublicKeys = await this.getDelegatePublicKeysSortedByVoteWeight(
+			tx,
+		);
+
+		// Delete delegate list and create new updated list
+		await this.storage.entities.RoundDelegates.delete(
+			{
+				round_gte: round,
```

Comment:
I'm not sure if we need to use round_gte because undo is already deleting the rounds that is not needed. What would be the case we would still have a round number with greater number.

I'd prefer to delete only round itself to see other problems could happen in dpos. I feel like this could suppress a bigger problem.

Reply:
You are right about this. We thought more about this as a safety net?

Reply:
I would say real safety is to see if we mess up in update function. If somebody breaks the functionality that takes care of deleting the list on undo, we may never notice it until we need to revert a block. In general I feel less safe when a function is doing another function's job. :)

Reply:
Changed `gte` to just the `round` number. :)